### ISSUE TITLE
[Proposition](Blade template compiler) Add indexing to foreach loops.

### ISSUE THREAD

**Issue by CzajekDC** [laravel/framework#12015]
## Disclaimer
I don't know that much about pull requests on Github. If it's irrelevant to post a request here, in issues, feel free to edit this comment or to write me about it.

## Proposition to Blade
The index in a foreach loop is a must have in my opinion. I always find myself doing this:
```php
<?php $i = 0; ?>
@foreach($array as $key)
    //
    {{ $i }}
    //
<?php $i++; ?>
@endforeach
```

It looks very messy, unelegant and how cool would it be to have this instead:
```php
@foreach($array as $key)
    //
    @index  or  {{ $index }}
    //
@endforeach
```

## Solution

Just change this ===> to this:
```php
@foreach($array as key)    ===>    <?php $index = 1; foreach ($array as $key): ?>
```
and this ===> to this:
```php
@endforeach    ===>    <?php $index++; endforeach; ?>
```

Then everyone could access the $index by saying:
```
@index or {{ $index }}
```

In `Illuminate\View\Compilers\BladeCompiler` ready compiler for `@index` method:
```php
/**
 * Compile blade custom index property inside a loop
 *
 * @param  string  $expression
 * @return string
 */
protected function compileIndex($expression)
{
    return '<?php echo $index; ?>';
}
```

## Conclusion
I think it's very easy to implement and that it is a must have for any templating language

**Comment by themsaid** [comment]
Why don't you just do:
```
@foreach($array as $index => $key)
    // Code inside the loop
@endforeach
```

**Comment by CzajekDC** [comment]
The thing is, imagine this example:
```php
$fields = [
    'company' => 'Company name',
    'address' => 'Address',
    'city' => 'City',
];

@foreach($fields as $field => $field_name)
    //
@endforeach
```

What then? You clearly need some index value.

You could do something like
```php
{{ array_search($field, array_keys($fields)) + 1 }}
```
or some other shit, but wouldn't it be nice to have an indexing number?

Twig has: http://twig.sensiolabs.org/doc/tags/for.html#the-loop-variable

**Comment by CupOfTea696** [comment]
You can use my (Counter)[https://github.com/CupOfTea696/Counter] package, just released the 1.0. Will write documentation at some point, but it will basically allow you to do this:

```html
@foreach(Counter::loop($array) as $key)
   @if (Counter::first())
      <!-- do something -->
   @endif
   {{ Counter::index() }}
@endforeach
```

For a full list of methods just check the main Counter.php file for the moment, it should be self-explanatory.

**Comment by CzajekDC** [comment]
+1 @CupOfTea696. Pretty cool.
